packages/network-controller/src/NetworkController.ts

```diff
@@ -991,201 +1293,627 @@ export class NetworkController extends BaseController<
+          network: rpcEndpoint.networkClientId,
+          ticker: nativeTokenName,
+          type: NetworkClientType.Infura,
         });
+      } else {
+        autoManagedNetworkClientRegistry[NetworkClientType.Custom][
+          rpcEndpoint.networkClientId
+        ] = createAutoManagedNetworkClient({
+          chainId,
+          rpcUrl: rpcEndpoint.url,
+          ticker: nativeTokenName,
+          type: NetworkClientType.Custom,
+        });
+      }
     }
 
+    const newNetworkConfiguration = {
+      ...fields,
+      rpcEndpoints: newRpcEndpoints,
+    };
+
     this.update((state) => {
-      state.networkConfigurations[upsertedNetworkConfigurationId] = {
-        id: upsertedNetworkConfigurationId,
-        ...sanitizedNetworkConfiguration,
-      };
+      state.networkConfigurationsByChainId[chainId] = newNetworkConfiguration;
     });
 
-    if (!existingNetworkConfiguration) {
-      this.#trackMetaMetricsEvent({
-        event: 'Custom Network Added',
-        category: 'Network',
-        referrer: {
-          url: referrer,
-        },
-        properties: {
-          chain_id: chainId,
-          symbol: ticker,
-          source,
-        },
-      });
-    }
+    this.#networkConfigurationsByNetworkClientId =
+      buildNetworkConfigurationsByNetworkClientId(
+        this.state.networkConfigurationsByChainId,
+      );
 
-    if (setActive) {
-      await this.setActiveNetwork(upsertedNetworkConfigurationId);
-    }
+    this.messagingSystem.publish(
+      `${controllerName}:networkAdded`,
+      newNetworkConfiguration,
+    );
 
-    return upsertedNetworkConfigurationId;
+    return newNetworkConfiguration;
   }
 
   /**
-   * Removes a custom network from state.
+   * Updates the configuration for a previously stored network filed under the
+   * given chain ID, creating and registering new network clients to represent
+   * RPC endpoints that have been added and destroying and unregistering
+   * existing network clients for RPC endpoints that have been removed.
    *
-   * This involves updating the `networkConfigurations` property in state as
-   * well and removing the network client that corresponds to the network from
-   * the client registry.
+   * Note that if `chainId` is changed, then all network clients associated with
+   * that chain will be removed and re-added, even if none of the RPC endpoints
+   * have changed.
    *
-   * @param networkConfigurationId - The ID of an existing network
-   * configuration.
+   * @param chainId - The chain ID associated with an existing network.
+   * @param fields - The object that describes the updates to the network/chain,
+   * including the new set of RPC endpoints which should front that chain.
+   * @returns The updated network configuration.
+   * @throws if `chainId` does not refer to an existing network configuration,
+   * or if any part of `fields` would produce invalid state.
+   * @see NetworkConfiguration
```

Comment:
nit:
```suggestion
   * @see {@link NetworkConfiguration}
```

Reply:
Fixed in [a600e92](https://github.com/MetaMask/core/pull/4286/commits/a600e9244be9f77c88e9369f6aab7c833b0374a1).